File: example-robots/java/SmartPathfindingRobot.java
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * Smart Pathfinding Robot - A demonstration of advanced pathfinding techniques
 * 
 * This robot shows how to implement proper pathfinding algorithms that consider:
 * 1. A* algorithm for optimal path finding
 * 2. Height differences and fuel costs
 * 3. Obstacle avoidance
 * 4. Fallback strategies
 * 
 * This is an educational example showing why simple greedy approaches 
 * (like "always go left/down") are suboptimal for RoboRush.
 */
public class SmartPathfindingRobot {
    
    private static final String ROBORUSH_URL = "http://localhost:8080/api";
    // Replace this token value with your own api token.
    private static final String API_TOKEN = "";
    
    private final HttpClient client = HttpClient.newHttpClient();
    private final ObjectMapper mapper = new ObjectMapper();
    
    private boolean isAttending = false;
    private final Map<Position, TileInfo> knownMap = new ConcurrentHashMap<>();
    private Position currentPosition;
    private Position targetPosition;
    private MapSize mapSize;
    private int fuel = 0;
    private int maxFuel = 0;
    
    /**
     * Position class to represent coordinates
     */
    public static class Position {
        public final int x, y;
        
        public Position(int x, int y) {
            this.x = x;
            this.y = y;
        }
        
        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (!(obj instanceof Position)) return false;
            Position pos = (Position) obj;
            return x == pos.x && y == pos.y;
        }
        
        @Override
        public int hashCode() {
            return Objects.hash(x, y);
        }
        
        @Override
        public String toString() {
            return String.format("Position(%d, %d)", x, y);
        }
        
        public int manhattanDistanceTo(Position other) {
            return Math.abs(this.x - other.x) + Math.abs(this.y - other.y);
        }
        
        public double euclideanDistanceTo(Position other) {
            return Math.sqrt(Math.pow(this.x - other.x, 2) + Math.pow(this.y - other.y, 2));
        }
        
        public List<Position> getNeighbors() {
            return Arrays.asList(
                new Position(x + 1, y), // EAST
                new Position(x - 1, y), // WEST
                new Position(x, y + 1), // SOUTH
                new Position(x, y - 1)  // NORTH
            );
        }
    }
    
    /**
     * Tile information from the map
     */
    public static class TileInfo {
        public final int height;
        public final String type;
        
        public TileInfo(int height, String type) {
            this.height = height;
            this.type = type;
        }
    }
    
    /**
     * Map size information
     */
    public static class MapSize {
        public final int width, height;
        
        public MapSize(int width, int height) {
            this.width = width;
            this.height = height;
        }
    }
    
    /**
     * A* pathfinding node
     */
    private static class PathNode implements Comparable<PathNode> {
        public final Position position;
        public final double gScore; // Cost from start
        public final double fScore; // Estimated total cost
        public final PathNode parent;
        
        public PathNode(Position position, double gScore, double fScore, PathNode parent) {
            this.position = position;
            this.gScore = gScore;
            this.fScore = fScore;
            this.parent = parent;
        }
        
        @Override
        public int compareTo(PathNode other) {
            return Double.compare(this.fScore, other.fScore);
        }
    }
    
    /**
     * Main robot loop
     */
    public void run() throws URISyntaxException, IOException, InterruptedException {
        System.out.println("Smart Pathfinding Robot booting up...");
        System.out.println("This robot demonstrates advanced pathfinding techniques");
        System.out.println("and shows why simple greedy approaches are suboptimal.");
        
        while (true) {
            try {
                JsonNode gameInfo = getGameInfo();
                String gameState = gameInfo.get("currentState").asText();
                
                switch (gameState) {
                    case "WAIT_FOR_PLAYERS":
                        if (!isAttending) {
                            isAttending = sendRobotAttending();
                        }
                        break;
                        
                    case "WAIT_FOR_ACTION":
                        if (isAttending) {
                            updateRobotState();
                            makeSmartMove();
                        }
                        break;
                        
                    case "ENDED":
                        isAttending = false;
                        System.out.println("Game ended. Resetting for next game.");
                        break;
                }
                
                Thread.sleep(250);
                
            } catch (Exception e) {
                System.err.println("Error in main loop: " + e.getMessage());
                Thread.sleep(1000);
            }
        }
    }
    
    /**
     * Get general game information
     */
    private JsonNode getGameInfo() throws URISyntaxException, IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(new URI(ROBORUSH_URL + "/game"))
                .GET()
                .build();
        
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        JsonNode gameInfo = mapper.readTree(response.body());
        
        // Update map size and target if available
        if (gameInfo.has("mapSize")) {
            JsonNode mapSizeNode = gameInfo.get("mapSize");
            this.mapSize = new MapSize(
                mapSizeNode.get("width").asInt(),
                mapSizeNode.get("height").asInt()
            );
        }
        
        if (gameInfo.has("targetPosition") && !gameInfo.get("targetPosition").isNull()) {
            JsonNode targetNode = gameInfo.get("targetPosition");
            this.targetPosition = new Position(
                targetNode.get("x").asInt(),
                targetNode.get("y").asInt()
            );
        }
        
        return gameInfo;
    }
    
    /**
     * Register robot for the current game
     */
    private boolean sendRobotAttending() throws URISyntaxException, IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(new URI(ROBORUSH_URL + "/robot/attend"))
                .header("Authorization", "Bearer " + API_TOKEN)
                .POST(HttpRequest.BodyPublishers.ofString(""))
                .build();
        
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        boolean success = response.statusCode() == 200;
        System.out.println("Attend the current game: " + success);
        return success;
    }
    
    /**
     * Update robot's current state and known map
     */
    private void updateRobotState() throws URISyntaxException, IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(new URI(ROBORUSH_URL + "/robot"))
                .header("Authorization", "Bearer " + API_TOKEN)
                .GET()
                .build();
        
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() == 200) {
            JsonNode robotData = mapper.readTree(response.body());
            
            // Update position and fuel
            JsonNode posNode = robotData.get("position");
            this.currentPosition = new Position(posNode.get("x").asInt(), posNode.get("y").asInt());
            this.fuel = robotData.get("fuel").asInt();
            this.maxFuel = robotData.get("maxFuel").asInt();
            
            // Update known map with scan results
            if (robotData.has("lastResult") && !robotData.get("lastResult").isNull()) {
                updateKnownMap(robotData.get("lastResult"));
            }
        }
    }
    
    /**
     * Update the known map with scan results
     */
    private void updateKnownMap(JsonNode scanResult) {
        if (scanResult.isArray()) {
            for (JsonNode tile : scanResult) {
                JsonNode posNode = tile.get("position");
                Position pos = new Position(posNode.get("x").asInt(), posNode.get("y").asInt());
                TileInfo tileInfo = new TileInfo(
                    tile.get("height").asInt(),
                    tile.get("type").asText()
                );
                knownMap.put(pos, tileInfo);
            }
        }
    }
    
    /**
     * Make an intelligent move using pathfinding
     */
    private void makeSmartMove() throws URISyntaxException, IOException, InterruptedException {
        if (currentPosition == null || targetPosition == null) {
            System.out.println("Missing position information for pathfinding");
            sendFallbackMove();
            return;
        }
        
        // Use A* pathfinding to find optimal route
        List<Position> path = findOptimalPath();
        
        if (path != null && path.size() > 1) {
            Position nextPosition = path.get(1); // First position is current position
            String direction = getDirectionTo(currentPosition, nextPosition);
            
            if (direction != null) {
                System.out.println("Smart pathfinding: Moving " + direction + " towards target");
                sendMoveAction(direction);
            } else {
                System.out.println("Could not determine direction from pathfinding");
                sendFallbackMove();
            }
        } else {
            System.out.println("No path found with A* algorithm, using fallback");
            sendFallbackMove();
        }
    }
    
    /**
     * Implement A* pathfinding algorithm considering fuel costs and height differences
     */
    private List<Position> findOptimalPath() {
        if (currentPosition == null || targetPosition == null) {
            return null;
        }
        
        PriorityQueue<PathNode> openSet = new PriorityQueue<>();
        Set<Position> closedSet = new HashSet<>();
        Map<Position, PathNode> nodeMap = new HashMap<>();
        
        PathNode startNode = new PathNode(
            currentPosition,
            0,
            heuristic(currentPosition, targetPosition),
            null
        );
        
        openSet.add(startNode);
        nodeMap.put(currentPosition, startNode);
        
        while (!openSet.isEmpty()) {
            PathNode current = openSet.poll();
            
            if (current.position.equals(targetPosition)) {
                // Reconstruct path
                return reconstructPath(current);
            }
            
            closedSet.add(current.position);
            
            // Check all neighbors
            for (Position neighborPos : getValidNeighbors(current.position)) {
                if (closedSet.contains(neighborPos)) {
                    continue;
                }
                
                double moveCost = getMoveCost(current.position, neighborPos);
                double tentativeGScore = current.gScore + moveCost;
                
                if (!nodeMap.containsKey(neighborPos)) {
                    // New node
                    PathNode neighborNode = new PathNode(
                        neighborPos,
                        tentativeGScore,
                        tentativeGScore + heuristic(neighborPos, targetPosition),
                        current
                    );
                    nodeMap.put(neighborPos, neighborNode);
                    openSet.add(neighborNode);
                    
                } else if (tentativeGScore < nodeMap.get(neighborPos).gScore) {
                    // Better path found
                    PathNode existingNode = nodeMap.get(neighborPos);
                    openSet.remove(existingNode); // Remove old node
                    
                    PathNode improvedNode = new PathNode(
                        neighborPos,
                        tentativeGScore,
                        tentativeGScore + heuristic(neighborPos, targetPosition),
                        current
                    );
                    nodeMap.put(neighborPos, improvedNode);
                    openSet.add(improvedNode);
                }
            }
        }
        
        return null; // No path found
    }
    
    /**
     * Reconstruct the path from the goal node
     */
    private List<Position> reconstructPath(PathNode node) {
        List<Position> path = new ArrayList<>();
        PathNode current = node;
        
        while (current != null) {
            path.add(current.position);
            current = current.parent;
        }
        
        Collections.reverse(path);
        return path;
    }
    
    /**
     * Get valid neighboring positions
     */
    private List<Position> getValidNeighbors(Position position) {
        List<Position> neighbors = new ArrayList<>();
        
        for (Position neighbor : position.getNeighbors()) {
            // Filter out positions outside map bounds
            if (mapSize != null && 
                (neighbor.x < 0 || neighbor.x >= mapSize.width || 
                 neighbor.y < 0 || neighbor.y >= mapSize.height)) {
                continue;
            }
            neighbors.add(neighbor);
        }
        
        return neighbors;
    }
    
    /**
     * Calculate the cost to move between two positions considering height differences
     */
    private double getMoveCost(Position from, Position to) {
        double baseCost = 1.0;
        
        // If we know the height information, factor it in
        TileInfo fromTile = knownMap.get(from);
        TileInfo toTile = knownMap.get(to);
        
        if (fromTile != null && toTile != null) {
            int heightDiff = toTile.height - fromTile.height;
            // Going uphill costs more fuel
            if (heightDiff > 0) {
                baseCost += heightDiff;
            }
        }
        
        return baseCost;
    }
    
    /**
     * Heuristic function for A* algorithm
     */
    private double heuristic(Position from, Position to) {
        // Euclidean distance with slight preference for straight lines
        double euclidean = from.euclideanDistanceTo(to);
        double manhattan = from.manhattanDistanceTo(to);
        
        return euclidean + 0.001 * manhattan;
    }
    
    /**
     * Get the direction string to move from one position to another
     */
    private String getDirectionTo(Position from, Position to) {
        int dx = to.x - from.x;
        int dy = to.y - from.y;
        
        if (dx == 1 && dy == 0) return "EAST";
        if (dx == -1 && dy == 0) return "WEST";
        if (dx == 0 && dy == 1) return "SOUTH";
        if (dx == 0 && dy == -1) return "NORTH";
        
        return null;
    }
    
    /**
     * Send a move action to the API
     */
    private void sendMoveAction(String direction) throws URISyntaxException, IOException, InterruptedException {
        String json = String.format("{\"direction\": \"%s\"}", direction);
        
        HttpRequest request = HttpRequest.newBuilder(new URI(ROBORUSH_URL + "/robot/action/move"))
                .header("Authorization", "Bearer " + API_TOKEN)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() == 200) {
            System.out.println("Successfully moved " + direction);
        } else {
            try {
                JsonNode error = mapper.readTree(response.body());
                System.out.println("Move failed: " + error.get("status") + " - " + error.get("message"));
            } catch (Exception e) {
                System.out.println("Move failed with status code: " + response.statusCode());
            }
        }
    }
    
    /**
     * Fallback move strategy when pathfinding fails
     */
    private void sendFallbackMove() throws URISyntaxException, IOException, InterruptedException {
        if (currentPosition == null || targetPosition == null) {
            sendWaitAction();
            return;
        }
        
        // Improved fallback: consider distance and cost
        List<String> directions = Arrays.asList("EAST", "WEST", "SOUTH", "NORTH");
        List<Position> positions = Arrays.asList(
            new Position(currentPosition.x + 1, currentPosition.y),
            new Position(currentPosition.x - 1, currentPosition.y),
            new Position(currentPosition.x, currentPosition.y + 1),
            new Position(currentPosition.x, currentPosition.y - 1)
        );
        
        String bestDirection = null;
        int bestDistance = Integer.MAX_VALUE;
        double bestCost = Double.MAX_VALUE;
        
        for (int i = 0; i < directions.size(); i++) {
            Position pos = positions.get(i);
            
            // Check if position is valid
            if (mapSize != null && 
                (pos.x < 0 || pos.x >= mapSize.width || pos.y < 0 || pos.y >= mapSize.height)) {
                continue;
            }
            
            int distance = pos.manhattanDistanceTo(targetPosition);
            double cost = getMoveCost(currentPosition, pos);
            
            // Choose based on distance first, then cost
            if (distance < bestDistance || (distance == bestDistance && cost < bestCost)) {
                bestDirection = directions.get(i);
                bestDistance = distance;
                bestCost = cost;
            }
        }
        
        if (bestDirection != null) {
            System.out.println("Fallback move: " + bestDirection);
            sendMoveAction(bestDirection);
        } else {
            System.out.println("No valid moves available, waiting");
            sendWaitAction();
        }
    }
    
    /**
     * Send a wait action when no move is possible
     */
    private void sendWaitAction() throws URISyntaxException, IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(new URI(ROBORUSH_URL + "/robot/action/wait"))
                .header("Authorization", "Bearer " + API_TOKEN)
                .POST(HttpRequest.BodyPublishers.ofString(""))
                .build();
        
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() == 200) {
            System.out.println("Waiting this turn");
        } else {
            System.out.println("Failed to send wait action");
        }
    }
    
    /**
     * Main entry point for the Smart Pathfinding Robot.
     * 
     * This example demonstrates:
     * 1. A* pathfinding algorithm implementation
     * 2. Considering height differences and fuel costs  
     * 3. Proper fallback strategies
     * 4. Educational comments explaining the approach
     * 
     * Key learning points:
     * - Simple "go left/down" strategies are suboptimal
     * - Height differences significantly affect fuel consumption
     * - Proper pathfinding considers obstacles and costs
     * - Fallback strategies are important for robustness
     */
    public static void main(String[] args) throws URISyntaxException, IOException, InterruptedException {
        new SmartPathfindingRobot().run();
    }
}

File: example-robots/python/SmartPathfindingRobot.py
import logging
import math
from time import sleep
from typing import List, Dict, Tuple, Optional, Set
import heapq

import requests

# This is the url to the RoboRush host.
ROBORUSH_URL = "http://localhost:8080/api"

# Replace this token value with your own api token.
# The token can be retrieved in the profile settings on the web ui.
API_TOKEN = ""

class Position:
    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y
    
    def __eq__(self, other):
        return self.x == other.x and self.y == other.y
    
    def __hash__(self):
        return hash((self.x, self.y))
    
    def __repr__(self):
        return f"Position({self.x}, {self.y})"
    
    def distance_to(self, other: 'Position') -> int:
        """Manhattan distance"""
        return abs(self.x - other.x) + abs(self.y - other.y)
    
    def euclidean_distance_to(self, other: 'Position') -> float:
        """Euclidean distance for heuristic"""
        return math.sqrt((self.x - other.x)**2 + (self.y - other.y)**2)

class PathNode:
    def __init__(self, position: Position, g_score: float, f_score: float, parent: Optional['PathNode'] = None):
        self.position = position
        self.g_score = g_score  # Cost from start to this node
        self.f_score = f_score  # Estimated total cost through this node
        self.parent = parent
    
    def __lt__(self, other):
        return self.f_score < other.f_score

class SmartPathfindingRobot:
    """
    A sophisticated robot that demonstrates proper pathfinding techniques.
    
    This robot shows how to:
    1. Use A* algorithm for optimal pathfinding
    2. Consider height differences and fuel costs
    3. Handle obstacles and blocked positions
    4. Implement fallback strategies
    
    This is an example of how players should approach pathfinding instead of 
    using simple greedy approaches like "always go left/down".
    """
    
    def __init__(self):
        self.is_attending = False
        self.known_map = {}
        self.map_size = None
        self.target_position = None
        self.current_position = None
        self.fuel = 0
        self.max_fuel = 0
        
    def main(self):
        logging.basicConfig(level=logging.INFO)
        logging.info("Smart Pathfinding Robot booting up...")
        logging.info("This robot demonstrates advanced pathfinding techniques")
        logging.info("and shows why simple greedy approaches are suboptimal.")
        
        while True:
            try:
                current_game_info = self.get_game_infos()
                
                if not self.is_attending and current_game_info["currentState"] == "WAIT_FOR_PLAYERS":
                    self.is_attending = self.send_robot_attending()
                    
                elif self.is_attending and current_game_info["currentState"] == "WAIT_FOR_ACTION":
                    self.update_robot_state()
                    self.make_smart_move()
                    
                elif current_game_info["currentState"] == "ENDED":
                    self.is_attending = False
                    logging.info("Game ended. Resetting for next game.")
                    
                sleep(0.25)
                
            except Exception as e:
                logging.error(f"Error in main loop: {e}")
                sleep(1)
    
    def get_game_infos(self) -> Dict:
        """Get general game information"""
        response = requests.get(ROBORUSH_URL + "/game")
        game_info = response.json()
        
        # Update map size and target if available
        if "mapSize" in game_info:
            self.map_size = game_info["mapSize"]
        if "targetPosition" in game_info and game_info["targetPosition"]:
            self.target_position = Position(
                game_info["targetPosition"]["x"],
                game_info["targetPosition"]["y"]
            )
            
        return game_info
    
    def send_robot_attending(self) -> bool:
        """Register robot for the current game"""
        response = requests.post(
            ROBORUSH_URL + "/robot/attend", 
            data="", 
            headers={"Authorization": "Bearer " + API_TOKEN}
        )
        
        success = response.status_code == 200
        logging.info(f"Attend the current game: {success}")
        return success
    
    def update_robot_state(self):
        """Update robot's current state and known map"""
        try:
            response = requests.get(
                ROBORUSH_URL + "/robot", 
                headers={"Authorization": "Bearer " + API_TOKEN}
            )
            
            if response.status_code == 200:
                robot_data = response.json()
                self.current_position = Position(
                    robot_data["position"]["x"],
                    robot_data["position"]["y"]
                )
                self.fuel = robot_data["fuel"]
                self.max_fuel = robot_data["maxFuel"]
                
                # Update known map with scan results
                if "lastResult" in robot_data and robot_data["lastResult"]:
                    self.update_known_map(robot_data["lastResult"])
                    
        except Exception as e:
            logging.error(f"Error updating robot state: {e}")
    
    def update_known_map(self, scan_result):
        """Update the known map with scan results"""
        if isinstance(scan_result, list):
            for tile in scan_result:
                pos = Position(tile["position"]["x"], tile["position"]["y"])
                self.known_map[pos] = {
                    "height": tile["height"],
                    "type": tile["type"]
                }
    
    def make_smart_move(self):
        """Make an intelligent move using pathfinding"""
        if not self.current_position or not self.target_position:
            logging.warning("Missing position information for pathfinding")
            self.send_fallback_move()
            return
        
        # Use A* pathfinding to find optimal route
        path = self.find_optimal_path()
        
        if path and len(path) > 1:
            next_position = path[1]  # First position in path is current position
            direction = self.get_direction_to(self.current_position, next_position)
            
            if direction:
                logging.info(f"Smart pathfinding: Moving {direction} towards target")
                self.send_move_action(direction)
            else:
                logging.warning("Could not determine direction from pathfinding")
                self.send_fallback_move()
        else:
            logging.warning("No path found with A* algorithm, using fallback")
            self.send_fallback_move()
    
    def find_optimal_path(self) -> Optional[List[Position]]:
        """
        Implement A* pathfinding algorithm considering fuel costs and height differences.
        This is a simplified version - in a real implementation, you'd want to 
        consider actual fuel costs based on height differences.
        """
        if not self.current_position or not self.target_position:
            return None
        
        open_set = []
        closed_set: Set[Position] = set()
        
        start_node = PathNode(
            position=self.current_position,
            g_score=0,
            f_score=self.heuristic(self.current_position, self.target_position)
        )
        
        heapq.heappush(open_set, start_node)
        node_map = {self.current_position: start_node}
        
        while open_set:
            current = heapq.heappop(open_set)
            
            if current.position == self.target_position:
                # Reconstruct path
                return self.reconstruct_path(current)
            
            closed_set.add(current.position)
            
            # Check all neighbors
            for neighbor_pos in self.get_valid_neighbors(current.position):
                if neighbor_pos in closed_set:
                    continue
                
                move_cost = self.get_move_cost(current.position, neighbor_pos)
                tentative_g_score = current.g_score + move_cost
                
                if neighbor_pos not in node_map:
                    # New node
                    neighbor_node = PathNode(
                        position=neighbor_pos,
                        g_score=tentative_g_score,
                        f_score=tentative_g_score + self.heuristic(neighbor_pos, self.target_position),
                        parent=current
                    )
                    node_map[neighbor_pos] = neighbor_node
                    heapq.heappush(open_set, neighbor_node)
                    
                elif tentative_g_score < node_map[neighbor_pos].g_score:
                    # Better path found
                    neighbor_node = node_map[neighbor_pos]
                    neighbor_node.g_score = tentative_g_score
                    neighbor_node.f_score = tentative_g_score + self.heuristic(neighbor_pos, self.target_position)
                    neighbor_node.parent = current
        
        return None  # No path found
    
    def reconstruct_path(self, node: PathNode) -> List[Position]:
        """Reconstruct the path from the goal node"""
        path = []
        current = node
        
        while current:
            path.append(current.position)
            current = current.parent
        
        path.reverse()
        return path
    
    def get_valid_neighbors(self, position: Position) -> List[Position]:
        """Get valid neighboring positions"""
        neighbors = [
            Position(position.x + 1, position.y),  # EAST
            Position(position.x - 1, position.y),  # WEST
            Position(position.x, position.y + 1),  # SOUTH
            Position(position.x, position.y - 1),  # NORTH
        ]
        
        # Filter out positions outside map bounds
        if self.map_size:
            neighbors = [
                pos for pos in neighbors
                if 0 <= pos.x < self.map_size["width"] and 0 <= pos.y < self.map_size["height"]
            ]
        
        return neighbors
    
    def get_move_cost(self, from_pos: Position, to_pos: Position) -> float:
        """
        Calculate the cost to move between two positions.
        This considers height differences if known.
        """
        base_cost = 1.0
        
        # If we know the height information, factor it in
        if from_pos in self.known_map and to_pos in self.known_map:
            from_height = self.known_map[from_pos]["height"]
            to_height = self.known_map[to_pos]["height"]
            height_diff = to_height - from_height
            
            # Going uphill costs more fuel
            if height_diff > 0:
                base_cost += height_diff
        
        return base_cost
    
    def heuristic(self, from_pos: Position, to_pos: Position) -> float:
        """
        Heuristic function for A* algorithm.
        Uses Euclidean distance with slight preference for straight lines.
        """
        euclidean = from_pos.euclidean_distance_to(to_pos)
        manhattan = from_pos.distance_to(to_pos)
        
        # Slight preference for straight lines
        return euclidean + 0.001 * manhattan
    
    def get_direction_to(self, from_pos: Position, to_pos: Position) -> Optional[str]:
        """Get the direction string to move from one position to another"""
        dx = to_pos.x - from_pos.x
        dy = to_pos.y - from_pos.y
        
        if dx == 1 and dy == 0:
            return "EAST"
        elif dx == -1 and dy == 0:
            return "WEST"
        elif dx == 0 and dy == 1:
            return "SOUTH"
        elif dx == 0 and dy == -1:
            return "NORTH"
        else:
            return None
    
    def send_move_action(self, direction: str):
        """Send a move action to the API"""
        response = requests.post(
            ROBORUSH_URL + "/robot/action/move",
            json={"direction": direction},
            headers={"Authorization": "Bearer " + API_TOKEN}
        )
        
        if response.status_code == 200:
            logging.info(f"Successfully moved {direction}")
        else:
            try:
                error_json = response.json()
                logging.warning(f"Move failed: {error_json.get('status', 'Unknown')} - {error_json.get('message', 'No message')}")
            except:
                logging.warning(f"Move failed with status code: {response.status_code}")
    
    def send_fallback_move(self):
        """
        Fallback move strategy when pathfinding fails.
        Still better than pure greedy as it considers multiple factors.
        """
        if not self.current_position or not self.target_position:
            # If we don't know where we are or where to go, just wait
            self.send_wait_action()
            return
        
        # Simple but improved approach: consider distance and fuel
        possible_moves = [
            ("EAST", Position(self.current_position.x + 1, self.current_position.y)),
            ("WEST", Position(self.current_position.x - 1, self.current_position.y)),
            ("SOUTH", Position(self.current_position.x, self.current_position.y + 1)),
            ("NORTH", Position(self.current_position.x, self.current_position.y - 1))
        ]
        
        # Filter valid moves and sort by distance to target
        valid_moves = []
        for direction, pos in possible_moves:
            if (not self.map_size or 
                (0 <= pos.x < self.map_size["width"] and 0 <= pos.y < self.map_size["height"])):
                distance = pos.distance_to(self.target_position)
                cost = self.get_move_cost(self.current_position, pos)
                valid_moves.append((direction, distance, cost))
        
        if valid_moves:
            # Sort by distance (primary) and cost (secondary)
            valid_moves.sort(key=lambda x: (x[1], x[2]))
            best_direction = valid_moves[0][0]
            logging.info(f"Fallback move: {best_direction}")
            self.send_move_action(best_direction)
        else:
            logging.warning("No valid moves available, waiting")
            self.send_wait_action()
    
    def send_wait_action(self):
        """Send a wait action when no move is possible"""
        response = requests.post(
            ROBORUSH_URL + "/robot/action/wait",
            headers={"Authorization": "Bearer " + API_TOKEN}
        )
        
        if response.status_code == 200:
            logging.info("Waiting this turn")
        else:
            logging.warning("Failed to send wait action")

def main():
    """
    Main entry point for the Smart Pathfinding Robot.
    
    This example demonstrates:
    1. A* pathfinding algorithm implementation
    2. Considering height differences and fuel costs
    3. Proper fallback strategies
    4. Educational comments explaining the approach
    
    Key learning points:
    - Simple "go left/down" strategies are suboptimal
    - Height differences significantly affect fuel consumption
    - Proper pathfinding considers obstacles and costs
    - Fallback strategies are important for robustness
    """
    robot = SmartPathfindingRobot()
    robot.main()

if __name__ == '__main__':
    main()